lib/room.cpp

```diff
@@ -360,37 +360,95 @@ class Room::Private {
+            QJsonObject {
+                { QStringLiteral("major"), 1 },
+                { QStringLiteral("minor"), 0 } } }
+        };
+        {
+            QJsonArray sessionsJson;
+            for (const auto &session : groupSessions) {
+                auto pickleResult = session.second->pickle(connection->picklingMode());
+                sessionsJson += QJsonObject {
+                    {QStringLiteral("sender_key"), session.first.first},
+                    {QStringLiteral("session_id"), session.first.second},
+                    {QStringLiteral("pickle"), QString(pickleResult)}
+                };
+            }
+            rootObj.insert(QStringLiteral("sessions"), sessionsJson);
+        }
+
+#if QT_VERSION >= QT_VERSION_CHECK(5, 15, 0)
+        const auto data = QJsonDocument(rootObj).toJson(QJsonDocument::Compact);
+#else
+        QJsonDocument json { rootObj };
+        const auto data = json.toJson(QJsonDocument::Compact);
+#endif
+
+        outFile.write(data.data(), data.size());
+        qCDebug(E2EE) << "Megolm sessions saved to" << outFile.fileName();
+    }
     bool addInboundGroupSession(QString senderKey, QString sessionId,
                                 QString sessionKey)
     {
-        // new e2ee TODO:
-        /*
-        if (groupSessions.contains({ senderKey, sessionId })) {
-            qCDebug(E2EE) << "Inbound Megolm session" << sessionId
+        if (groupSessions.find(qMakePair(senderKey, sessionId)) != groupSessions.end()) {
+            qCWarning(E2EE) << "Inbound Megolm session" << sessionId
                           << "with senderKey" << senderKey << "already exists";
             return false;
         }
 
-        QOlmInboundGroupSession* megolmSession;
-        try {
-            megolmSession = new QOlmInboundGroupSession(sessionKey.toLatin1(),
-                                                    InboundGroupSession::Init,
-                                                    q);
-        } catch (QOlmError* e) {
-            qCDebug(E2EE) << "Unable to create new InboundGroupSession"
-                          << e->what();
-            return false;
-        }
-        if (megolmSession->id() != sessionId) {
-            qCDebug(E2EE) << "Session ID mismatch in m.room_key event sent "
-                             "from sender with key"
-                          << senderKey;
+        std::unique_ptr<QOlmInboundGroupSession> megolmSession = QOlmInboundGroupSession::create(sessionKey.toLatin1());
+        if (megolmSession->sessionId() != sessionId) {
+            qCWarning(E2EE) << "Session ID mismatch in m.room_key event sent "
+                             "from sender with key" << senderKey;
             return false;
         }
-        groupSessions.insert({ senderKey, sessionId }, megolmSession);
-        */
+        qCWarning(E2EE) << "Adding inbound session";
+        groupSessions[qMakePair(senderKey, sessionId)] = std::move(megolmSession);
```

Comment:
```suggestion
        groupSessions[{senderKey, sessionId}] = std::move(megolmSession);
```
should be enough, no?